### PR TITLE
Add http streaming api documentation

### PR DESCRIPTION
**Pull request by SimonWoolf**
Initial rough-and-ready version of SSE / HTTP streaming api documentation.

### CODE REVIEW

**Comment by SimonWoolf** [comment]
@mattheworiordan tagged you as this is a good time if you have any feedback about the format, key names etc.

File: content/streaming-api/index.textile [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+h3(#event-stream). Plain event stream
+
+h6. GET realtime.ably.io/event-stream
+
+Start a streaming HTTP request.
+
+h5. Request parameters
+
+- channels := (mandatory) one or more channel names, separated by commas (or the @separator@ if specified). Non-url-safe characters should be URL-encoded (for example, @?channels=foo%3Fbar@ will subscribe to the channel @foo?bar@). Alias: @channel@
+- separator := (optional) a separator, to enable easy subscriptions to channels with commas in their name. For example, @?channels=?separator=|&channel=fo,o|ba,r@ will subcribe to the two channels @fo,o@ and @ba,r@.
```
Comment (mattheworiordan) [inline content/streaming-api/index.textile]: Typo:

` @?channels=?separator=|&channel=fo,o|ba,r@`

should be the following I think

`@?separator=|&channel=fo,o|ba,r@`

File: content/streaming-api/index.textile [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+- v := (mandatory) the version of the api you are requesting. The current (and only) version of the API is 1.1, so only @v=1.1@ is accepted
+- key := (optional) an Ably API key to use, if using basic auth
+- accessToken := (optional) an Ably auth token to use, if using token auth
```
Comment (mattheworiordan) [inline content/streaming-api/index.textile]: Can this not be `token`?  Would JWT work, I assume so?

File: content/streaming-api/index.textile [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+- lastEvent := (optional) an @id@ to resume from, see "connection state recovery":#recovery below
+- enveloped := (optional, default true) if true, the @data@ from each event envelope for a @message@ event will be a "Message":/realtime/types#message object. If false, it will be the payload from the message directly. See "Envelope format":#envelope-format below
+- heartbeats := (optional, default false) if @true@ will use an explicit heartbeat event rather than a newline as a keepalive packet
```
Comment (mattheworiordan) [inline content/streaming-api/index.textile]: I thought we were going to support fields so that subscribers could select only the fields they want sent.

File: content/streaming-api/index.textile [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+h5(#envelope-format). Envelope format
+
+Once a streaming response is established, every line (other than empty lines sent as keepalive packets) will be a simple JSON object of the following form:
+
+bc[json]. {
+  event: <string, the event type, such as "message", "presence", "error", or "heartbeat">,
+  data: <string, the message data, see below. Not present for some event types,
+         e.g. heartbeats>,
+  id: <string, the id to use to resume from this point, see 'connection state
+       recovery' below. Only presence for message and presence events>
```
Comment (mattheworiordan) [inline content/streaming-api/index.textile]: Only **present**

File: content/streaming-api/index.textile (hunk 3 above)
Comment (SimonWoolf) [reply content/streaming-api/index.textile]: Yeah, I just didn't do it in the MVP (wasn't essential).  I'll add it at some point, maybe as part of the work I do next week

File: content/streaming-api/index.textile (hunk 2 above)
Comment (SimonWoolf) [reply content/streaming-api/index.textile]: Auth is processed by the same code (AuthMgr) whether for rest, websocket, or this. So JWT will work fine, and we can add `token` as an alias of `accessToken` (and `access_token`) if we want as long as nothing else uses that param (and I don't imagine anything does), sure

Comment (mattheworiordan) [reply content/streaming-api/index.textile]: > So JWT will work fine, and we can add token as an alias of accessToken

Ok, sounds good.  Given we have `key` as opposed to `accessKey`, I think using `token` is preferable

**Comment by tomczoink** [comment]
We should update our adapters page https://www.ably.io/adapters once this is merged.

File: content/sse/index.textile [hunk 5]
```diff
@@ -0,0 +1,119 @@
+---
+title: SSE and Raw HTTP Streaming API
+section: streaming-api
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Does this appear in the nav anywhere?  I can't see it in https://ably-docs-pr-666.herokuapp.com/

File: content/sse/index.textile [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  API Routes:
+    - Server-sent events#sse
+    - Plain event stream#event-stream
+---
+
+The Ably SSE and raw HTTP streaming API provides a way to get a realtime stream of events from Ably in circumstances where using a full Ably Realtime client library, or even an MQTT library, is impractical.
+
+HTTP streaming allows for a request from a client to be held by a server, allowing it to push data to the client without further requests. This, much like WebSockets, help avoid the overhead involved in normal HTTP requests. "Server-sent events":https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events (SSE) is another method of achieving this for web pages, usually done through the use of "EventSource":https://developer.mozilla.org/en-US/docs/Web/API/EventSource.
+
+This feature is currently in beta. If you're interested in using this feature, or have suggestions on how we can improve it, please "get in touch":https://www.ably.io/contact.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Perhaps:

This feature is currently in beta, however is ready for production loads. If you're interested in using this feature, or have suggestions on how we can improve it, we would like you to "get in touch":https://www.ably.io/contact.

Comment (mattheworiordan) [reply content/sse/index.textile]: I also think this should appear in a note i.e. on the right.  I don't think it belongs in the middle of the content.

File: content/sse/index.textile [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+It is subscribe-only: you can not interact with the channel, including to publish, enter presence, query the presence set, attach and detach from channels (without closing and re-opening the stream), or anything else.
+
+Where possible, we strongly recommend the use of one of our "Realtime client libraries":https://www.ably.io/download, which provide "more features and higher reliability":https://support.ably.io/solution/articles/3000061848, and the full use of our "normal realtime messaging API":/realtime.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Should we not say something like.  

"Customers who do not want to use a client library on platforms that support SSE, and only require simple subscribe-only streams, may choose to use SSE because it's an open standard, simple, and requires no SDKs on the client-side.  However, where possible, we strongly recommend..."

File: content/sse/index.textile [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+h2(#authentication). Authentication
+
+The same authentication options as used in the REST API also apply here. So for basic auth, a querystring parameter of @key@ or an @Authorization: Basic <base64-encoded key>@ header; for token auth, a querystring parameter of @accessToken@ containing an access token or an @Authorization: Bearer <base64-encoded token>@ header. See "REST API authentication":/rest-api#authentication for more information.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: I don't think we should be recommending that customers deal with authentication via our raw REST API.  I think we should say that customers typically use a token that is issued from their servers using an Ably SDK and an API key.  It is also possible to simply use an API, however for security reasons that is rarely advisable. Find out why token authentication is preferred in almost any case.  In addition, Ably supports JWT tokens which an be included in the @token@ query param.  I would recommend we then link to a JSBin with examples of each please.  ALso, we decided on using @token@ as opposed to @accessToken@ I believe.

Comment (mattheworiordan) [inline content/sse/index.textile]: > The same authentication options as used in the REST API also apply here

Surely it's the same authentication options used by any clients connecting to Ably over any protocol apply here.  

Then we should link to find out more about Ably authentication, not linking to the raw REST API where customers are going to have to try and figure out how to generate tokens themselves, which is almost never correct.

File: content/sse/index.textile [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Extra line break

File: content/sse/index.textile [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+h2(#connection-state-recovery). Connection state recovery
+
+If you are disconnected, you can supply the @id@ of the last event you received as the @lastEvent@ querystring parameter of a new connection, to resume from that point.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: This does not apply to SSE, at least in that SSE automatically does this.  This is very misleading as it makes a developer think they need to do this themselves.  This only applies when using XHR

File: content/sse/index.textile [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+Connection state is only retained for two minutes. See "Connection state explained":/realtime/connection#connection-state-explained for full documentation.
+
+Due to this feature being in beta, it is currently subject to other restrictions. For example, at the moment you can not resume from part-way through a backlog of messages that was being sent as part of a previous resume; you can only resume from a message (so you can't resume unless you've received at least one message on the connection); and you can only resume from the id of a message that was sent in the last two minutes (so it will not work if the last message sent on the connection was more than two minutes ago). We expect to fix all of these by the time this feature graduates out of beta.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Perhaps add "Get in touch" if you'd like to find out when these shortcomings will be fixed in producvtion.

File: content/sse/index.textile (hunk 7 above)
Comment (mattheworiordan) [reply content/sse/index.textile]: I think at this point, we should tell people that SSE is incredibly simple, open, and here is a code sample showing how easy it is.  With a link to JSBin to try it out live.

File: content/sse/index.textile [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+The @/event-stream@ endpoint will give an SSE response if the @Accept@ header is set to @text/event-stream@. The @/sse@ endpoint is also provided as an easier way of forcing an SSE response.
+
+h5. Request parameters
+
+Identical to the request parameters for a plain event stream above.
+
+h5(#envelope-format). Envelope format
+
+See for a plain event stream above, except instead of a JSON object with @id@, @event@, @data@ members, you get an SSE event.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: I don't follow this.  What do people nee to see above?

File: content/sse/index.textile [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+Keepalive packets are sent as sse comments (@:keepalive@).
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Is this relevant to end users?  Not one for me to comment on, but do SSE clients even emit comments?

File: content/sse/index.textile [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+h6. GET realtime.ably.io/sse
+
+Start a streaming HTTP request that conforms to the "Server-Sent Events":https://www.w3.org/TR/eventsource/ spec, for ease of consuming with an SSE library.
+
+The @/event-stream@ endpoint will give an SSE response if the @Accept@ header is set to @text/event-stream@. The @/sse@ endpoint is also provided as an easier way of forcing an SSE response.
+
+h5. Request parameters
+
+Identical to the request parameters for a plain event stream above.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: There is no plain event stream above?  Also, why should a developer need to look at another reference.  Just show the arguments here.  Let our customers be lazy, we should do the hard work of making these docs readable and easy to sue.

File: content/sse/index.textile [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+h5(#envelope-format). Envelope format
+
+See for a plain event stream above, except instead of a JSON object with @id@, @event@, @data@ members, you get an SSE event.
+
+Keepalive packets are sent as sse comments (@:keepalive@).
+
+h5. Example
+
+```[sh]
+curl "https://rest.ably.io/sse?key={{API KEY}}&channel=example&v=1.1"
```
Comment (mattheworiordan) [inline content/sse/index.textile]: This is a curl example, but people will be using clients and not accessing raw streams.  Please provide examples in other languages.

File: content/sse/index.textile [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+Start a streaming HTTP request.
+
+h5. Request parameters
+
+- channels := (mandatory) one or more channel names, separated by commas (or the @separator@ if specified). Non-url-safe characters should be URL-encoded (for example, @?channels=foo%3Fbar@ will subscribe to the channel @foo?bar@). Alias: @channel@.
+- separator := (optional) a separator, to enable easy subscriptions to channels with commas in their name. For example, @?separator=|&channel=fo,o|ba,r@ will subcribe to the two channels @fo,o@ and @ba,r@.
+- v := (mandatory) the version of the api you are requesting. The current (and only) version of the API is 1.1, so only @v=1.1@ is accepted.
+- key := (optional) an Ably API key to use, if using basic auth.
+- accessToken := (optional) an Ably auth token to use, if using token auth.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: I think we agreed to use the shortcut @token@

File: content/sse/index.textile [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+h6. GET realtime.ably.io/event-stream
+
+Start a streaming HTTP request.
+
+h5. Request parameters
+
+- channels := (mandatory) one or more channel names, separated by commas (or the @separator@ if specified). Non-url-safe characters should be URL-encoded (for example, @?channels=foo%3Fbar@ will subscribe to the channel @foo?bar@). Alias: @channel@.
+- separator := (optional) a separator, to enable easy subscriptions to channels with commas in their name. For example, @?separator=|&channel=fo,o|ba,r@ will subcribe to the two channels @fo,o@ and @ba,r@.
+- v := (mandatory) the version of the api you are requesting. The current (and only) version of the API is 1.1, so only @v=1.1@ is accepted.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: Could we have mandatory stuff at the top please

File: content/sse/index.textile [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+- key := (optional) an Ably API key to use, if using basic auth.
+- accessToken := (optional) an Ably auth token to use, if using token auth.
+- lastEvent := (optional) an @id@ to resume from, see "connection state recovery":#recovery below.
```
Comment (mattheworiordan) [inline content/sse/index.textile]: @SimonWoolf do we additionally support `rewind` now?

Comment (mattheworiordan) [reply content/sse/index.textile]: Is `lastEvent` supported for SSE, or more accurately, should we tell people it's not needed except when starting a new SSE connection that resumes from a previous connection?

Comment (mattheworiordan) [reply content/sse/index.textile]: This recovery link does not exist, and refers to a link below, which does not exist (it's above).

File: content/sse/index.textile [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+If @enveloped@ is false, the @data@ will for a @message@ or @presence@ event be the @data@ payload from the @Message@ or @PresenceMessage@. Other events are unaffected.
+
+Non-string payloads will be encoded as before, but without enveloping you will not have the benefit of the @encoding@ field to tell you what encoding has been done.
+
+Note that failures on opening the connection (for example, invalid authentication details) may be sent as a non-streamed http response (with a response body of the form @{"error": <ErrorInfo>}@), not an @error@ event in a streamed response.
+
+h5. Example
+
+bc[sh]. curl "https://rest.ably.io/event-stream?key={{API KEY}}&channel=example&v=1.1"
```
Comment (mattheworiordan) [inline content/sse/index.textile]: We could easily provide a trivial example in a JSBin and inline that subscribes via XHR, and resumes automatically setting `lastEvent`

File: layouts/_nav.html.erb [hunk 20]
```diff
@@ -4,6 +4,7 @@
   <%= nav_items 'realtime', :title => 'Realtime client lib' %>
   <%= nav_items 'rest', :title => 'REST client lib' %>
   <%= nav_items 'rest-api', :title => 'REST API' %>
+  <%= nav_items 'sse', :title => 'SSE' %>
```
Comment (mattheworiordan) [inline layouts/_nav.html.erb]: I don't think this is appearing.

File: content/sse/index.textile (hunk 8 above)
Comment (SimonWoolf) [reply content/sse/index.textile]: > Surely it's the same authentication options used by any clients connecting to Ably over any protocol apply here. Then we should link to find out more about Ably authentication

I don't follow you here. It's the same underlying protocol as our client libraries use sure, but that's not much help for someone who isn't a client library developer... Our normal authentication docs are written assuming you're using a client library, the only places the actual protocol is documented are the client library development guide, and the raw REST api docs

(Unless you're just saying we should copy & paste (or better transclude) the relevant bits from the rest api docs here, rather than linking?)

> , not linking to the raw REST API where customers are going to have to try and figure out how to generate tokens themselves, which is almost never correct.

Nothing about using the raw REST api means customers have to generate tokens by hand. The [rest api docs](https://www.ably.io/documentation/rest-api#token-authentication) just explain, once you've obtained a token (from a client lib on your auth server or otherwise), how to include it in the REST request you make so that we make (`Authorization` header or querystring param). Which is the the same as what someone using SSE or light-http-streaming will have to do

File: content/sse/index.textile (hunk 18 above)
Comment (SimonWoolf) [reply content/sse/index.textile]: > do we additionally support rewind now?

Yes (currently just with a number though)

> Is lastEvent supported for SSE, or more accurately, should we tell people it's not needed except when starting a new SSE connection that resumes from a previous connection?

sgtm (if supplied it takes precedence over the SSE `headers['last-event-id']`)

File: content/streaming-api/index.textile [hunk 21]
```diff
@@ -12,7 +12,9 @@ jump_to:
     - Server-sent events#sse
 ---
 
-The Ably raw HTTP streaming API provides a way to get a realtime stream of events from Ably in circumstances where using a full Ably Realtime client library, or even an MQTT library, is impractical.
+The Ably SSE + raw HTTP streaming API provides a way to get a realtime stream of events from Ably in circumstances where using a full Ably Realtime client library, or even an MQTT library, is impractical.
+
+HTTP streaming allows for a request from a client to be held by a server, allowing it to push data to the client without further requests. This, much like WebSockets, help avoid the overhead involved in normal HTTP requests. "Server-sent events":https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events (SSE) is another method of achieving this for web pages, usually done through the use of "EventSource":https://developer.mozilla.org/en-US/docs/Web/API/EventSource.
```
Comment (SimonWoolf) [inline content/streaming-api/index.textile]: > Server-sent events (SSE) is another method of achieving this for web pages

@tomasmcm It's not really a different method, it's a thin layer on top of HTTP streaming. It's not limited to web pages either, eg we have SSE python and elixir examples in the tutorials repo

**Comment by SimonWoolf** [comment]
@tomczoink do you want to update per latest round of feedback or would you like me to? (or any specific bits you want me to look at?)

**Comment by tomczoink** [comment]
I'm happy to implement the feedback @SimonWoolf , I'd appreciate you giving it a review once I'm done though.

**Comment by accesstoken** [comment]
Hi
I shouldn’t be involved in this conversation. Wrong person I think.

**Comment by mattheworiordan** [comment]
Sorry @accesstoken, see https://github.com/ably/docs/pull/666#discussion_r292225722, should have wrapped that in a `code block`!

File: content/sse/index.textile (hunk 18 above)
Comment (tomczoink) [reply content/sse/index.textile]: How is `rewind` to be defined @SimonWoolf ?

**Comment by tomczoink** [comment]
@mattheworiordan @SimonWoolf I've implemented changes, a review on current state would be much appreciated.

Also, for the point:

> We don't explain why developers would use the raw HTTP Streaming API, and we should. Why would they use that over SSE?

What would be a reason to use HTTP Streaming over SSE? SSE not being available on the platform, or any other points to make?

**Comment by SimonWoolf** [comment]
> What would be a reason to use HTTP Streaming over SSE? SSE not being available on the platform

basically that yup, if you have to do it from scratch yourself then parsing json is probs easier than implementing an sse client

File: content/sse/index.textile (hunk 18 above)
Comment (SimonWoolf) [reply content/sse/index.textile]: `- rewind := (optional integer) if specified, will send a backlog of this many messages once the connection opens. For example, @rewind=1@ will give you the most recent message sent on the channel. This is best-effort — only messages from within the last two minutes will be available, and only if the channel has been continuously active since the message was sent; it is not a replacement for the "history API":/rest/history. It only has an effect for new connections; when "resuming a previous connection":#connection-state-recovery, it is ignored in favour of sending you the messages you missed since you were last connected`

**Comment by mattheworiordan** [comment]
I've made quite a few small changes to these docs, please review individual commits and notes.

@tomczoink @remotekieran @SimonWoolf can you review please?

File: content/sse/index.textile [hunk 22]
```diff
@@ -140,15 +140,15 @@ Start a streaming HTTP request, for ease of consuming with any HTTP library that
...
+- lastEvent := **optional**. An @id@ to resume from. Only required when starting a new SSE connection which resumes from a previous connection.
+- rewind := **optional**. An integer which, if specified, will send a backlog of the number of messages specified once the connection opens. For example, @rewind=1@ will give you the most recent message sent on the channel. This is best-effort — only messages from within the last two minutes will be available, and only if the channel has been continuously active since the message was sent; it is not a replacement for the "history API":/rest/history. It only has an effect for new connections; when resuming a previous connection using @lastEvent@, it is ignored in favour of sending you the messages you missed since you were last connected.
+- enveloped := **optional**. Default is **true**. If true, the @data@ from each event envelope for a @message@ event will be a "Message":/realtime/types#message object. If false, it will be the payload from the message directly. See "Envelope format":#envelope-format below.
+- heartbeats := ≈ if @true@ will use an explicit heartbeat event rather than a newline as a keepalive packet.
```
Comment (SimonWoolf) [inline content/sse/index.textile]: @tomczoink `≈`?

**Comment by mattheworiordan** [comment]
@tomczoink FYI, I added `curl` as a new code type, so we need to make sure this works in the website when this is merged in.  It may not, I can't recall how the language mappings are shared (or not) between the website & the docs repo.

File: content/sse/index.textile (hunk 22 above)
Comment (mattheworiordan) [inline content/sse/index.textile]: Why is this one not flagged as optional?

Comment (tomczoink) [reply content/sse/index.textile]: Fixed with https://github.com/ably/docs/pull/666/commits/2e53135582c4897a94af7a2cc966669c7dbf20c8.

**Comment by mattheworiordan** [comment]
Minor request, I noticed in https://www.ably.io/tutorials/sse-and-http-streaming, that we refer to url incorrectly, it's an acronym, `URL`.  I notice we also sometimes refer to mqtt in lower case too.  Mind just doing a grep to fix those?

**Comment by mattheworiordan** [comment]
@tomczoink don't worry about it, done in dc7e784

File: content/streaming-api/index.textile (hunk 2 above)
Comment (paddybyers) [reply content/streaming-api/index.textile]: Except that "access token" is widely known and accepted terminology, among the various other token types used in oauth schemes (client token, user token, refresh token etc). I don't see the advantage in adding an ambiguous `token` param.

Comment (mattheworiordan) [reply content/streaming-api/index.textile]: >  I don't see the advantage in adding an ambiguous token param.

Brevity and consistency with our client SDKs which only refer to tokens as `token` or `tokenDetails`

Comment (paddybyers) [reply content/streaming-api/index.textile]: So if we have both `accessToken` and `token`, which is the preferred or primary name, and which is the alias? In other situations we've argued against having multiple ways of doing the same thing, because it only creates confusion.

Comment (mattheworiordan) [reply content/streaming-api/index.textile]: > In other situations we've argued against having multiple ways of doing the same thing, because it only creates confusion.

My preference is just `token`
